lib/srv/discovery/access_graph_azure.go

```diff
@@ -0,0 +1,409 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discovery
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/entitlements"
+	accessgraphv1alpha "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha"
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/discovery/fetchers/azuresync"
+)
+
+// reconcileAccessGraphAzure fetches Azure resources, creates a set of resources to delete and upsert based on
+// the previous fetch, and then sends the delete and upsert results to the Access Graph stream
+func (s *Server) reconcileAccessGraphAzure(
+	ctx context.Context,
+	currentTAGResources *azuresync.Resources,
+	stream accessgraphv1alpha.AccessGraphService_AzureEventsStreamClient,
+	features azuresync.Features,
+) error {
+	type fetcherResult struct {
+		result *azuresync.Resources
+		err    error
+	}
+
+	// Get all the fetchers
+	allFetchers := s.getAllTAGSyncAzureFetchers()
+	if len(allFetchers) == 0 {
+		// If there are no fetchers, we don't need to continue.
+		// We will send a delete request for all resources and return.
+		upsert, toDel := azuresync.ReconcileResults(currentTAGResources, &azuresync.Resources{})
+
+		if err := azurePush(stream, upsert, toDel); err != nil {
+			s.Log.ErrorContext(ctx, "Error pushing empty resources to TAGs", "error", err)
+		}
+		return trace.Wrap(errNoAccessGraphFetchers)
+	}
+
+	// Fetch results concurrently
+	resultsC := make(chan fetcherResult, len(allFetchers))
+	tokens := make(chan struct{}, 3)
+	accountIds := map[string]struct{}{}
+	for _, fetcher := range allFetchers {
+		fetcher := fetcher
+		accountIds[fetcher.GetSubscriptionID()] = struct{}{}
+		tokens <- struct{}{}
+		go func() {
+			defer func() {
+				<-tokens
+			}()
```

Comment:
Is this an implementation of something similar to `sync.WaitGroup`?

Reply:
yes but sync.Waitgroup doesn't allow you to limit the number of concurrent calls